modules/tinymce/src/core/main/ts/caret/CaretUtils.ts

```diff
@@ -76,14 +76,10 @@ const findNode = (node: Node, direction: number, predicateFn: (node: Node) => bo
   return null;
 };
 
-const getEditingHost = (node: Node, rootNode?: Node): Node | undefined => {
-  for (node = node.parentNode; node && node !== rootNode; node = node.parentNode) {
-    if (isContentEditableTrue(node)) {
-      return node;
-    }
-  }
-
-  return rootNode;
+const getEditingHost = (node: Node, rootNode: HTMLElement): HTMLElement => {
+  return PredicateFind.ancestor(SugarElement.fromDom(node), (node) => isContentEditableTrue(node.dom), (node) => node.dom === rootNode)
+    .map((elm) => elm.dom)
+    .getOr(rootNode) as HTMLElement;
 };
```

Comment:
```suggestion
const getEditingHost = (node: Node, rootNode: HTMLElement): HTMLElement =>
  PredicateFind.ancestor(SugarElement.fromDom(node), (node): node is SugarElement<HTMLElement> => isContentEditableTrue(node.dom), (node) => node.dom === rootNode)
    .map((elm) => elm.dom)
    .getOr(rootNode);
```
This avoids the cast which we should be avoiding if possible. While this still does do some casting, it only applies for the predicate within the `ancestor` lookup. It could also be worthwhile splitting the predicate out into a separate function 🤷  For example:
```ts
const getEditingHost = (node: Node, rootNode: HTMLElement): HTMLElement => {
  const isCETrue = (node: SugarElement<Node>): node is SugarElement<HTMLElement> => isContentEditableTrue(node.dom);
  const isRoot = (node: SugarElement<Node>) => node.dom === rootNode;
  return PredicateFind.ancestor(SugarElement.fromDom(node), isCETrue, isRoot)
    .map((elm) => elm.dom)
    .getOr(rootNode);
};
```

Reply:
Right yeah it would cast anyway not sure that can be avoided here since we know that it is a HTMLElement because of the predicate so I was unsure where to stick that casting. I updated the function to your suggestions. I wonder how much effort it would be to remove the SugarElement wrapping object it serves little to no purpose in a typed language and it would make things like this where you try to marry sugar with regular dom more easy. In rockfest we use a sugar like library but with regular nodes, elements etc. Anyway, thanks for the suggestion.

Reply:
We use unwrapped references in RTC because it's an abstract type. Arguably sugar wrappers serve a lot less purpose now that they're not thunked, but they do at least (mostly) enforce a 'not null' requirement.

Reply:
But wouldn't strict mode eventually give us that for ts too or is that only solving implicit anys not enforcing foo | null on everything.

Reply:
The not null benefit happens through our `fromDom` method. Otherwise we're relying on the TS libdom to declare when DOM methods might return null. Maybe strict mode will help with this? We're a very long way from enabling strict in TinyMCE though.

Reply:
Right yeah that makes sense. I guess my brain has been destroyed by not having to deal with null values for a couple of years. :smile: